crates/ra_ide/src/completion/complete_attribute.rs

```diff
@@ -0,0 +1,582 @@
+        should_be_inner: false,
+    },
+    AttrCompletion { label: "link", snippet: None, should_be_inner: false },
+    AttrCompletion { label: "macro_export", snippet: None, should_be_inner: false },
+    AttrCompletion { label: "macro_use", snippet: None, should_be_inner: false },
+    AttrCompletion {
+        label: "must_use",
+        snippet: Some(r#"must_use = "${0:reason}""#),
+        should_be_inner: false,
+    },
+    AttrCompletion { label: "no_mangle", snippet: None, should_be_inner: false },
+    AttrCompletion { label: "no_std", snippet: None, should_be_inner: true },
+    AttrCompletion { label: "non_exhaustive", snippet: None, should_be_inner: false },
+    AttrCompletion { label: "panic_handler", snippet: None, should_be_inner: true },
+    AttrCompletion { label: "path", snippet: Some("path =\"${0:path}\""), should_be_inner: false },
+    AttrCompletion { label: "proc_macro", snippet: None, should_be_inner: false },
+    AttrCompletion { label: "proc_macro_attribute", snippet: None, should_be_inner: false },
+    AttrCompletion {
+        label: "proc_macro_derive",
+        snippet: Some("proc_macro_derive(${0:Trait})"),
+        should_be_inner: false,
+    },
+    AttrCompletion {
+        label: "recursion_limit",
+        snippet: Some("recursion_limit = ${0:128}"),
+        should_be_inner: true,
+    },
+    AttrCompletion { label: "repr", snippet: Some("repr(${0:C})"), should_be_inner: false },
+    AttrCompletion {
+        label: "should_panic",
+        snippet: Some(r#"expected = "${0:reason}""#),
+        should_be_inner: false,
+    },
+    AttrCompletion {
+        label: "target_feature",
+        snippet: Some("target_feature = \"${0:feature}\""),
+        should_be_inner: false,
+    },
+    AttrCompletion { label: "test", snippet: None, should_be_inner: false },
+    AttrCompletion { label: "used", snippet: None, should_be_inner: false },
+    AttrCompletion { label: "warn", snippet: Some("warn(${0:lint})"), should_be_inner: false },
+    AttrCompletion {
+        label: "windows_subsystem",
+        snippet: Some(r#"windows_subsystem = "${0:subsystem}""#),
+        should_be_inner: true,
+    },
+];
+
+struct AttrCompletion {
+    label: &'static str,
+    snippet: Option<&'static str>,
+    should_be_inner: bool,
+}
+
+pub(super) fn complete_attribute(acc: &mut Completions, ctx: &CompletionContext) {
+    if !ctx.is_attribute {
+        return;
+    }
+
+    let is_inner = ctx
+        .original_token
+        .ancestors()
+        .find_map(Attr::cast)
+        .map(|attr| attr.kind() == AttrKind::Inner)
+        .unwrap_or(false);
+
+    for attr_completion in ATTRIBUTES {
+        let mut item = CompletionItem::new(
+            CompletionKind::Attribute,
+            ctx.source_range(),
+            attr_completion.label,
+        )
+        .kind(CompletionItemKind::Attribute);
+        if let Some(snippet) = attr_completion.snippet {
+            item = item.insert_snippet(snippet);
```

Comment:
This needs a rebase, we've changed API around snippets

Reply:
Oops, done